ekf: fix optical flow bugs

ChristophTobler opened the pull request:
Fixes https://github.com/PX4/Firmware/issues/8102

The lines
```
optflow_sample_new.flowRadXYcomp(0) = optflow_sample_new.flowRadXY(0) - optflow_sample_new.gyroXYZ(0);
optflow_sample_new.flowRadXYcomp(1) = optflow_sample_new.flowRadXY(1) - optflow_sample_new.gyroXYZ(1);
```
result in,
```
optflow_sample_new.flowRadXYcomp(0) = + flow->gyrodata(0) - (- flow->gyrodata(0))
optflow_sample_new.flowRadXYcomp(1) = + flow->gyrodata(1) - (- flow->gyrodata(1))
```
if the vehicle is on ground and flow quality is not good, which is non-zero.

ChristophTobler commented:
I added a check to see if the terrain is valid. Otherwise it shouldn't use flow.

priseborough commented:
I now know why the flow is being rejected on the ground. With very bad quality,  the integration interval is going to zero which is stopping flow measurements being pushed to the buffer.

priseborough commented on EKF/control.cpp:
```diff
@@ -383,8 +383,8 @@ void Ekf::controlOpticalFlowFusion()
-		// fuse the data
-		if (_control_status.flags.opt_flow) {
+		// fuse the data if the terrain/distance to bottom is valid
+		if (_control_status.flags.opt_flow && get_terrain_valid()) {
```
This is going to cause problems on ground when the range finder is out of range

ChristophTobler replied:
We could do something like
```
// use rng_gnd_clearance if on ground
				if (!_ekf.get_in_air_status()) {
					range_finder.current_distance = _params->rng_gnd_clearance;

				// check if distance sensor is within working boundarie
				} else if (range_finder.min_distance >= range_finder.current_distance ||
					         range_finder.max_distance <= range_finder.current_distance) {
					range_finder_updated = false;
				}
```
in the Firmware. But then handheld tests wouldn't work.

priseborough replied:
I was thinking that if on ground, an absence of range data for longer than TBD or range data below min or above some reasonable max value for hand carriage would cause rng_gnd_clearance to be used.

ChristophTobler replied:
or actually this way you can still do handheld tests
```
// check if distance sensor is within working boundaries
if (range_finder.min_distance >= range_finder.current_distance ||
	range_finder.max_distance <= range_finder.current_distance) {
	// use rng_gnd_clearance if on ground
	if (_ekf.get_in_air_status()) {
		range_finder_updated = false;

	} else {
		range_finder.current_distance = _params->rng_gnd_clearance;
	}
}
```

ChristophTobler commented:
I fixed a small bug which made the flow predictions go unstable. I also added the PX4Flow behavior to the simulation https://github.com/PX4/sitl_gazebo/pull/132 (that also revealed the bug)

ChristophTobler commented:
Works well on my hardware
![px4_flow_ekf_issue_fixed](https://user-images.githubusercontent.com/14769636/31444033-4c70fd76-ae9b-11e7-878a-9de8bdefc2b4.png)

I guess we can get this in as CI is broken?

priseborough commented:
I've fixed up the reversion to constant position mode which makes it more robust to handling  when on ground.

priseborough commented:
SITL testing is showing large innovation spikes on attitude changes, but this is also happening on master.

![screen shot 2017-10-12 at 9 07 22 am](https://user-images.githubusercontent.com/3596952/31469261-dee61862-af2c-11e7-8a21-cc273cef4c1b.png)

ChristophTobler commented:
Cool, thanks!
